dotnet/src/SemanticKernel/Diagnostics/Verify.cs

```diff
@@ -58,7 +58,7 @@ internal static void ValidFunctionName([NotNull] string? functionName)
     internal static void ValidFunctionParamName([NotNull] string? functionParamName)
     {
         NotEmpty(functionParamName, "The function parameter name cannot be empty");
-        Regex pattern = new("^[0-9A-Za-z_]*$");
+        Regex pattern = new("^[0-9A-Za-z_-]*$");
```

Comment:
we cannot change this without changing also the template language and the documentation. Overall I wouldn't add new chars without a design conversation. Unless there's a requirement, could you undo this?

Reply:
![image](https://user-images.githubusercontent.com/68852919/229594983-2b5e028a-5977-4c46-802b-0e07fdb4bd95.png)
The change was introduced to add support for parameters that have dashes in their names like - profile-key, X-Api-Version. The parameters with dashes are extensively used in template expressions in OpenAPI documents. 

We can add a workaround to replace all dashes in names by underscores and preserve original name in a separate property so that we can use original name if it's used as parameter name for a query string parameter, as a request header name or a body property name.

On the other hand, how difficult is it to support dashes in parameters for the template language?

Reply:
to change the template language we need to consider
* impact on real variables used in C# and python
* how params are extracted from C# and python
* integration and unit tests
* documentation
* which other chars should we add? e.g. query params allow $ and OData uses `$` extensively, but `$` has a different meaning in the template

I would suggest looking for an approach that doesn't affect the template language, decoupling the two.

Reply:
Dashes are common across REST API args. What alternative could there be? I think we have to support this.

It's up to the language hosting the function to define what args it supports. This would work in C# and Python too, since these map to values in a dictionary -- not method args. If we do one day dynamically map native method args to SKContext parameters, we'll do it the same way as System.Text.Json -- with additional attributes and/or automatic transforms.

Reply:
@SergeyMenshykh has brought in an alternative to keep the parameter name parsing regex unchanged for now. Let's have a more relaxed design conversation after this comes in.